### PR TITLE
std: Deprecate std::old_io::fs

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit deprecates the majority of std::old_io::fs in favor of std::fs and
its new functionality. Some functions remain non-deprecated but are now behind a
feature gate called `old_fs`. These functions will be deprecated once
suitable replacements have been implemented.

The compiler has been migrated to new `std::fs` and `std::path` APIs where
appropriate as part of this change.

[breaking-change]

### CODE REVIEW

**Comment by huonw** [comment]
Travis is failing.

**Comment by alexcrichton** [comment]
Ah oops, should be taken care of now

File: mk/docs.mk [hunk 1]
```diff
@@ -273,11 +273,13 @@ compiler-docs: $(COMPILER_DOC_TARGETS)
 trpl: doc/book/index.html
 
 doc/book/index.html: $(RUSTBOOK_EXE) $(wildcard $(S)/src/doc/trpl/*.md) | doc/
+	@$(call E, rustbook: $@)
```
Comment (aturon) [inline mk/docs.mk]: Is this intended? Seems unrelated to the goal of the PR...

File: src/librustc_trans/back/link.rs [hunk 2]
```diff
@@ -613,28 +616,32 @@ fn link_rlib<'a>(sess: &'a Session,
                 let bc_deflated_filename = obj_filename.with_extension(
                     &format!("{}.bytecode.deflate", i));
 
-                let bc_data = match fs::File::open(&bc_filename).read_to_end() {
-                    Ok(buffer) => buffer,
+                let mut bc_data = Vec::new();
+                match fs::File::open(&bc_filename).and_then(|mut f| {
```
Comment (aturon) [inline src/librustc_trans/back/link.rs]: These days I like to use `if let` for stuff like this. (No big deal obviously)

File: src/rustbook/error.rs [hunk 3]
```diff
@@ -21,63 +19,17 @@ pub type CliResult<T> = Result<T, CliError>;
 pub type CommandError = Box<Error + 'static>;
 pub type CommandResult<T> = Result<T, CommandError>;
 
-pub trait Error {
```
Comment (aturon) [inline src/rustbook/error.rs]: Ha!

**Comment by aturon** [comment]
LGTM

File: mk/docs.mk (hunk 1 above)
Comment (alexcrichton) [inline mk/docs.mk]: Ah yes I figured I'd just clean up the output here to be a little more consistent with the rest of the build (now that it doesn't print the chapters to stdout by default)

File: src/librustc_trans/back/link.rs (hunk 2 above)
Comment (alexcrichton) [inline src/librustc_trans/back/link.rs]: Ah yes I forgot about that, but with the new signature of `read_to_end` I suspect that may be more ergonomic.

**Comment by Manishearth** [comment]
Doctest! :smile: 

```
C:/bot/slave/auto-win-64-nopt-t/build/src/librustdoc\html\render.rs:500:36: 500:57 error: mismatched types:
 expected `&std::old_path::windows::Path`,
    found `&std::path::PathBuf`
(expected struct `std::old_path::windows::Path`,
    found struct `std::path::PathBuf`) [E0308]
C:/bot/slave/auto-win-64-nopt-t/build/src/librustdoc\html\render.rs:500     let _lock = ::flock::Lock::new(&cx.dst.join(".lock"));
                               
```

**Comment by Manishearth** [comment]
```

---- [parse-fail] parse-fail/circular_modules_main.rs stdout ----
	
error: expected error: on line 12 not found: circular modules
status: exit code: 101
command: PATH="i686-pc-windows-gnu/stage2/bin;C:\bot\slave\auto-win-32-opt\build\obj\i686-pc-windows-gnu\stage2\bin;C:\program files (x86)\mingw-w64\i686-4.9.1-win32-dwarf-rt_v3-rev1\mingw32\bin;C:\msys64\usr\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\msys64\usr\bin;C:\python27;C:\python27\scripts;C:\program files (x86)\inno setup 5;C:\program files (x86)\CMake\bin" i686-pc-windows-gnu/stage2/bin/rustc.exe C:/bot/slave/auto-win-32-opt/build/src/test/parse-fail/circular_modules_main.rs -L i686-pc-windows-gnu/test/parse-fail/ --target=i686-pc-windows-gnu -L i686-pc-windows-gnu/test/parse-fail\circular_modules_main.stage2-i686-pc-windows-gnulibaux -C prefer-dynamic -o i686-pc-windows-gnu/test/parse-fail\circular_modules_main.stage2-i686-pc-windows-gnu.exe --cfg rtopt --cfg debug -O -L i686-pc-windows-gnu/rt
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
C:/bot/slave/auto-win-32-opt/build/src/test/parse-fail\circular_modules_main.rs:12:5: 12:27 error: circular modules: C:/bot/slave/auto-win-32-opt/build/src/test/parse-fail\circular_modules_hello.rs -> C:/bot/slave/auto-win-32-opt/build/src/test/parse-fail\circular_modules_main.rs -> C:/bot/slave/auto-win-32-opt/build/src/test/parse-fail\circular_modules_hello.rs
C:/bot/slave/auto-win-32-opt/build/src/test/parse-fail\circular_modules_main.rs:12 mod circular_modules_hello; //~ ERROR: circular modules
                             
```

...what

**Comment by Manishearth** [comment]
```
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-opt/build/src/compiletest/runtest.rs:957:9: 957:34 error: imports are not allowed after non-item statements [E0154]
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-opt/build/src/compiletest/runtest.rs:957         use std::ascii::AsciiExt;
```

**Comment by Manishearth** [comment]
```
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:237:42: 237:54 error: mismatched types:
 expected `&mut core::ops::FnMut() -> std::path::PathBuf`,
    found `[closure /Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:237:42: 237:54]`
(expected &-ptr,
    found closure) [E0308]
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:237             get_install_prefix_lib_path: || panic!(),
                                                                                                                                       ^~~~~~~~~~~~
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:238:30: 238:45 error: the type of this value must be known in this context
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:238             realpath: |p| Ok(p.to_path_buf())
                                                                                                                           ^~~~~~~~~~~~~~~
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:238:23: 238:46 error: mismatched types:
 expected `&mut for<'r> core::ops::FnMut(&'r std::path::Path) -> core::result::Result<std::path::PathBuf, std::io::error::Error>`,
    found `[closure /Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:238:23: 238:46]`
(expected &-ptr,
    found closure) [E0308]
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:238             realpath: |p| Ok(p.to_path_buf())
                                                                                       
```

**Comment by Manishearth** [comment]
```
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:204:46: 204:58 error: mismatched types:
 expected `&mut core::ops::FnMut() -> std::path::PathBuf`,
    found `[closure /Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:204:46: 204:58]`
(expected &-ptr,
    found closure) [E0308]
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:204                 get_install_prefix_lib_path: || panic!(),
                                                                                                                                           ^~~~~~~~~~~~
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:205:34: 205:49 error: the type of this value must be known in this context
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:205                 realpath: |p| Ok(p.to_path_buf())
                                                                                                                               ^~~~~~~~~~~~~~~
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:205:27: 205:50 error: mismatched types:
 expected `&mut for<'r> core::ops::FnMut(&'r std::path::Path) -> core::result::Result<std::path::PathBuf, std::io::error::Error>`,
    found `[closure /Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:205:27: 205:50]`
(expected &-ptr,
    found closure) [E0308]
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/librustc_back/rpath.rs:205                 realpath: |p| Ok(p.to_path_buf())
                                                                                                                        ^~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to 3 previous errors
make: *** [i686-apple-darwin/stage2/test/rustc_backtest-i686-apple-darwin] Error 101
```

**Comment by Manishearth** [comment]
Bunch of lldb debuginfo failures

```

thread '[debuginfo-lldb] debuginfo/shadowed-argument.rs' panicked at 'explicit panic', /Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/compiletest/runtest.rs:1483


---- [debuginfo-lldb] debuginfo/shadowed-variable.rs stdout ----
	NOTE: compiletest thinks it is using LLDB version 310

error: line not found in debugger output: [...]$0 = false
status: exit code: 0
command: "python" "/Users/rustbuild/src/rust-buildbot/slave/auto-mac-32-opt/build/src/etc/lldb_batchmode.py" "i686-apple-darwin/test/debuginfo-lldb/shadowed-variable.stage2-i686-apple-darwin" "i686-apple-darwin/test/debuginfo-lldb/shadowed-variable.debugger.script"
```

**Comment by eddyb** [comment]
I see no explanation for the changes in pretty-printing, i.e. writing to a buffer and then saving that buffer to disk (or printing it out).
I guess I can work with that in #23017, if I move some things around... but I find it rather odd.

**Comment by alexcrichton** [comment]
@eddyb oh sorry about that, I initially didn't want to migrate the pretty printer just yet to the new I/O traits so it was easiest to just throw it all into a `Vec` and then print that to a `std::fs::File` for example. I wound up having to migrate the pretty printer, though, and I just forgot to go back and fixup what previously changed. Please feel free to revert back to what it previously was though!
